File: C++/binary-tree-upside-down.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root, TreeNode* parent=NULL) {
        if(!root) return parent;
        TreeNode* new_root = upsideDownBinaryTree(root->left, root);
        root->left = parent ? parent->right : NULL;
        root->right = parent;
        return new_root;
    }
};

File: Python/binary-tree-upside-down.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {TreeNode}
    def upsideDownBinaryTree(self, root, parent=None):
        if not root: return parent
        new_root = self.upsideDownBinaryTree(root.left, root)
        root.left = parent.right if parent else None
        root.right = parent
        return new_root
